crates/sui-core/src/authority/authority_store_types.rs

```diff
@@ -0,0 +1,139 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use move_core_types::language_storage::StructTag;
+use serde::{Deserialize, Serialize};
+use serde_with::serde_as;
+use serde_with::Bytes;
+use std::convert::{From, TryFrom};
+use sui_types::base_types::{ObjectDigest, SequenceNumber, TransactionDigest};
+use sui_types::crypto::{sha3_hash, Signable};
+use sui_types::error::SuiError;
+use sui_types::move_package::MovePackage;
+use sui_types::object::{Data, MoveObject, Object, Owner};
+
+/// Forked version of [`sui_types::object::Object`]
+/// Used for efficient storing of move objects in the database
+#[derive(Eq, PartialEq, Debug, Clone, Deserialize, Serialize, Hash)]
+pub struct StoreObject {
+    pub data: StoreData,
+    pub owner: Owner,
+    pub previous_transaction: TransactionDigest,
+    pub storage_rebate: u64,
+}
+
+/// Forked version of [`sui_types::object::Data`]
+/// Adds extra enum value `MoveMetadata`, which represents a reference to an object stored separately
+#[derive(Eq, PartialEq, Debug, Clone, Deserialize, Serialize, Hash)]
+pub enum StoreData {
+    Move(MoveObject),
+    Package(MovePackage),
+    IndirectObject(IndirectObjectMetadata),
+}
+
+/// Metadata of stored moved object
+#[derive(Eq, PartialEq, Debug, Clone, Deserialize, Serialize, Hash)]
+pub struct IndirectObjectMetadata {
+    version: SequenceNumber,
+    pub digest: ObjectDigest,
+}
+
+/// Separately stored move object
+#[serde_as]
+#[derive(Eq, PartialEq, Debug, Clone, Deserialize, Serialize, Hash)]
+pub struct StoreMoveObject {
+    pub type_: StructTag,
+    has_public_transfer: bool,
+    #[serde_as(as = "Bytes")]
+    contents: Vec<u8>,
+    /// reference count of `MoveMetadata` that point to the same content
+    /// once it hits 0, the object gets deleted by a compaction job
+    ref_count: usize,
+}
+
+impl<W> Signable<W> for StoreMoveObject
+where
+    W: std::io::Write,
+{
+    fn write(&self, writer: &mut W) {
+        write!(writer, "StoreMoveObject::").expect("Hasher should not fail");
+        bcs::serialize_into(writer, &self).expect("Message serialization should not fail");
+    }
+}
+
+impl StoreMoveObject {
+    pub fn digest(&self) -> ObjectDigest {
+        ObjectDigest::new(sha3_hash(self))
+    }
+}
+
+pub struct StoreObjectPair(pub StoreObject, pub Option<StoreMoveObject>);
+
+impl From<Object> for StoreObjectPair {
+    fn from(object: Object) -> Self {
+        let mut indirect_object = None;
+
+        let data = match object.data {
+            Data::Package(package) => StoreData::Package(package),
+            Data::Move(move_obj) => {
+                // TODO: add real heuristic to decide if object needs to be stored indirectly
+                if false {
```

Comment:
May be we have a variable here with `cfg(test)` to enable this code path for tests for now